File: bms/go/geo/geo_s2_CellID_Intersects/geo_s2_CellID_Intersects.go
package test

func (ci geo_s2_CellID) Intersects(oci geo_s2_CellID) bool {
	return uint64(oci.RangeMin()) <= uint64(ci.RangeMax()) && uint64(oci.RangeMax()) >= uint64(ci.RangeMin())
}

func (ci geo_s2_CellID) RangeMin() geo_s2_CellID { return geo_s2_CellID(uint64(ci) - (ci.lsb() - 1)) }

func (ci geo_s2_CellID) lsb() uint64 { return uint64(ci) & -uint64(ci) }

func (ci geo_s2_CellID) RangeMax() geo_s2_CellID { return geo_s2_CellID(uint64(ci) + (ci.lsb() - 1)) }

type geo_s2_CellID uint64

File: translations_only/fail/go/geo/geo_s2_CellID_Intersects/mistral/Restart/geo_s2_CellID_Intersects.rs

use std::u64;

pub struct GeoS2CellID(u64);

impl GeoS2CellID {
    pub fn intersects(&self, oci: GeoS2CellID) -> bool {
        (oci.range_min().0 as u64) <= (self.range_max().0 as u64)
            && (oci.range_max().0 as u64) >= (self.range_min().0 as u64)
    }

    pub fn range_min(&self) -> GeoS2CellID {
        GeoS2CellID(self.0 - (self.lsb() - 1))
    }

    pub fn lsb(&self) -> u64 {
        self.0 & ! (self.0 - 1)
    }

    pub fn range_max(&self) -> GeoS2CellID {
        GeoS2CellID(self.0 + (self.lsb() - 1))
    }
}
